sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubConsumer.cs

```diff
@@ -237,5 +348,260 @@ protected EventHubConsumer()
+        ///
+        private async Task UnsubscribeFromChannelAsync(Guid identifier)
+        {
+            if ((ActiveChannels.TryRemove(identifier, out var unsubscribeChannel)) && (ActiveChannels.Count == 0))
+            {
+                await ChannelSyncRoot.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    // If the channel was the last active channel and publishing is still marked as active, take
+                    // the necessary steps to stop publishing.
+
+                    if ((_isPublishingActive) && (ActiveChannels.Count == 0))
+                    {
+                        _channelPublishingTokenSource?.Cancel();
+
+                        if (_channelPublishingTask != null)
+                        {
+                            try
+                            {
+                                await _channelPublishingTask.ConfigureAwait(false);
+                            }
+                            catch (TaskCanceledException)
+                            {
+                               // This is an expected scenario; no action is needed.
+                            }
+                        }
+
+                        _channelPublishingTokenSource?.Dispose();
+                        _isPublishingActive = false;
+                    }
+                }
+                finally
+                {
+                    ChannelSyncRoot.Release();
+                }
+            }
+
+            // Attempt to finalize the channel, signaling that no more writes should be
+            // expected to occur.
+
+            unsubscribeChannel?.Writer.TryComplete();
+        }
+
+        /// <summary>
+        ///   Begins the background process responsible for receiving from the partition
+        ///   and publishing to all subscribed channels.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to signal the request to cancel the background publishing.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private Task StartBackgroundChannelPublishingAsync(CancellationToken cancellationToken) =>
+            Task.Run(async () =>
+            {
+                var failedAttemptCount = 0;
+                var receivedItems = default(IEnumerable<EventData>);
+                var retryDelay = default(TimeSpan?);
+                var activeException = default(Exception);
+
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    try
+                    {
+                        // Receive items in batches and then write them to the subscribed channels.  The channels will naturally
+                        // block if they reach their maximum queue size, so there is no need to throttle publishing.
```

Comment:
One channel would block publishing for everyone, is this the intended behavior?

Reply:
It is.   Not because I like it, but because I really can't come up with a better approach.   Maybe you'll think of something that I didn't, but I see the options as:

- If we have a slow consumer, the rest are blocked until the slow one clears room in its channel.   _(current implementation)_

- Use unbounded channels and potentially have resource use grow unchecked.  _(wasn't a fan of the trade-off here; its unlikely but it makes me uncomfortable)_

- Drop events for consumers that are blocked.   _(the consumer has a single view into the infinite stream of events; it can't move forward and keep an eye on where it left off without opening a new set of AMQP links.   I could queue them up somewhere, but then we're back at the "unbounded channel" appraoch, more or less.)_

I don't see multiple iterators against the same consumer as a recommended pattern or the majority case;  there's a few situations where it could be useful and makes more sense than enforcing a "you can only create a single iterator at a time" scenario.

Reply:
/cc: @chrisjbarr